be/src/exec/pipeline/fragment_executor.cpp

```diff
@@ -187,13 +187,17 @@ Status FragmentExecutor::prepare(ExecEnv* exec_env, const TExecPlanFragmentParam
     const auto& pipelines = _fragment_ctx->pipelines();
     const size_t num_pipelines = pipelines.size();
     size_t driver_id = 0;
+    size_t num_root_drivers = 0;
     for (auto n = 0; n < num_pipelines; ++n) {
         const auto& pipeline = pipelines[n];
         // DOP(degree of parallelism) of Pipeline's SourceOperator determines the Pipeline's DOP.
         const auto degree_of_parallelism = pipeline->source_operator_factory()->degree_of_parallelism();
         LOG(INFO) << "Pipeline " << pipeline->to_readable_string() << " parallel=" << degree_of_parallelism
                   << " fragment_instance_id=" << print_id(params.fragment_instance_id);
-        const bool is_root = (n == num_pipelines - 1);
+        // This happens when we create new pipelines at the end of original pipeline
+        // Theoretically we should set original pipeline non-root.
+        // But here for simplicity we just ignore that.
+        const bool is_root = (n == num_pipelines - 1) || (pipeline->is_root());
```

Comment:
The condition is little confusing to me, and my suggestion is that
1. Mark last of the pipeline as root in method `PipelineBuilder::build`
2. Replace the condition here with `pipeline->is_root()` only